### PR TITLE
issue: 1234612 Fix SYN_RCVD via SYN received

### PR DESCRIPTION
**Pull request by serglelv**

### CODE REVIEW

**Comment by serglelv** [comment]
@liranoz12 Hi Liran, please review.

**Comment by serglelv** [comment]
bot:retest

File: src/vma/lwip/tcp.c [hunk 1]
```diff
@@ -82,6 +82,13 @@ void register_tcp_seg_free(tcp_seg_free_fn fn)
 }
 #endif
 
+tcp_set_error_staus_fn external_tcp_set_error_status;
+
+void register_tcp_set_error_status(tcp_set_error_staus_fn fn)
+{
+	external_tcp_set_error_status = fn;
+}
+
```
Comment (liranoz12) [inline src/vma/lwip/tcp.c]: Consider moving m_error_status into the pcb struct.
This way you will not need to add this callback.

File: src/vma/lwip/tcp_out.c [hunk 2]
```diff
@@ -1121,6 +1121,11 @@ tcp_output(struct tcp_pcb *pcb)
        #if TCP_OVERSIZE_DBGCHECK
          seg->oversize_left = 0;
        #endif /* TCP_OVERSIZE_DBGCHECK */
+       if ((pcb->in_flag_hndlg & TCP_SYN) && (get_tcp_state(pcb) & SYN_RCVD)) {
+         pcb->in_flag_hndlg = 0;
+         pcb->rcv_nxt = 1;
+         seg->tcphdr->seqno = htonl(htonl(seg->tcphdr->seqno)-1);
+       }
```
Comment (liranoz12) [inline src/vma/lwip/tcp_out.c]: This will probably affect the fast path, can we avoid this addition ?

File: src/vma/sock/sockinfo_tcp.cpp [hunk 3]
```diff
@@ -3719,6 +3720,31 @@ int sockinfo_tcp::getsockopt_offload(int __level, int __optname, void *__optval,
...
+				ti->tcpi_rcv_mss = m_pcb.mss;
+				ti->tcpi_advmss = m_pcb.advtsd_mss;
+				ti->tcpi_snd_cwnd = m_pcb.cwnd;
+				ti->tcpi_snd_ssthresh = m_pcb.ssthresh;
+				ti->tcpi_unacked = m_pcb.unacked->seqno;
+				ret = 0;
+			} else {
+				errno = EINVAL;
+			}
+			break;
```
Comment (liranoz12) [inline src/vma/sock/sockinfo_tcp.cpp]: Is this related to the fix? if not we need to split it into two commits.

File: src/vma/lwip/tcp_in.c [hunk 4]
```diff
@@ -599,6 +599,16 @@ tcp_process(struct tcp_pcb *pcb, tcp_in_data* in_data)
...
+    /* received SYN? SYN_RCVD via SYN */
+    else if (in_data->flags & TCP_SYN) {
+      err = tcp_enqueue_flags(pcb, TCP_SYN);
+      if (err == ERR_OK) {
+        /* Changed the pcbs state now */
+        set_tcp_state(pcb, SYN_RCVD);
+        pcb->in_flag_hndlg |= TCP_SYN;
+        TCP_SET_ERROR_STATUS(pcb, 0);
+      }
+    }
```
Comment (liranoz12) [inline src/vma/lwip/tcp_in.c]: 1. Is this addition supposed to handle this case ?
https://tools.ietf.org/html/rfc793#section-3.4 
The "three-way handshake" is the procedure used to establish a
  connection.  This procedure normally is initiated by one TCP and
  responded to by another TCP.  The procedure also works if two TCP
  simultaneously initiate the procedure.  When simultaneous attempt
  occurs, each TCP receives a "SYN" segment which carries no
  acknowledgment after it has sent a "SYN". 

2. tcp_enqueue_flags is used to enqueue TCP options for transmission but i dont see any tcp_out() call here ?
Where we send the data ?

3. set_tcp_state(pcb, SYN_RCVD) is called only from tcp_listen_input() while with this change it can also be called from tcp_process during connect() - is this valid ?

File: src/vma/sock/sockinfo_tcp.cpp (hunk 3 above)
Comment (serglelv) [reply src/vma/sock/sockinfo_tcp.cpp]: This code related to the fix but only used in PD.  I agree this code should be removed.

File: src/vma/lwip/tcp_out.c (hunk 2 above)
Comment (serglelv) [reply src/vma/lwip/tcp_out.c]: This functionality probably can be moved to the slowtmr/fasttmr? I will review it.

File: src/vma/lwip/tcp_in.c (hunk 4 above)
Comment (serglelv) [reply src/vma/lwip/tcp_in.c]: 1. It seems yes. 2. The fast path is used. Review -wip. 3. tcp connect during connect  review - wip.

File: src/vma/lwip/tcp.c (hunk 1 above)
Comment (serglelv) [reply src/vma/lwip/tcp.c]: I agree. The code will be updated.
